# edwardbirdlab/BALROG-MON: Citations

## BALROG-MON
[Code Repository](https://github.com/edwardbirdlab/BALROG-MON)

> If using BALROG-MON please cite the [GitHub page](https://github.com/edwardbirdlab/BALROG-MON) and/or the publication below. Also, feel free to contact me (edwardbirdlab@gmail.com | edwardbird@ksu.edu) to let me know!

> 

## Nextflow
[Code Repository](https://github.com/nextflow-io/nextflow) | [Webpages](https://www.nextflow.io/)

> Di Tommaso, P., Chatzou, M., Floden, E. W., Barja, P. P., Palumbo, E., & Notredame, C. (2017). Nextflow enables reproducible computational workflows. _Nature Biotechnology, 35_(4), 316-319. https://doi.org/10.1038/nbt.3820

## Pipeline Tools
- **AMRFinderPlus**

    [Code Repository](https://github.com/ncbi/amr) | [Webpages](https://www.ncbi.nlm.nih.gov/pathogens/antimicrobial-resistance/AMRFinder/)

    > Feldgarden, M., Brover, V., Gonzalez-Escalona, N., Frye, J. G., Haendiges, J., Haft, D. H., ... & Klimke, W. (2021). AMRFinderPlus and the Reference Gene Catalog facilitate examination of the genomic links among antimicrobial resistance, stress response, and virulence. _Scientific Reports, 11_(1), 12728. https://doi.org/10.1038/s41598-021-91456-0

- **Bracken**

    [Code Repository](https://github.com/jenniferlu717/Bracken) | [Webpages](https://ccb.jhu.edu/software/bracken/)

    > Lu, J., Breitwieser, F. P., Thielen, P., & Salzberg, S. L. (2017). Bracken: estimating species abundance in metagenomics data. _PeerJ Computer Science, 3_, e104. https://doi.org/10.7717/peerj-cs.104

- **BUSCO: Benchmarking Universal Single-Copy Orthologue toolkit**

   [Code Repository](https://gitlab.com/ezlab/busco) | [Webpages](https://busco.ezlab.org/)

  > Manni, M., Berkeley, M. R., Seppey, M., Simão, F. A., & Zdobnov, E. M. (2021). BUSCO update: novel and streamlined workflows along with broader and deeper phylogenetic coverage for scoring of eukaryotic, prokaryotic, and viral genomes. _Molecular Biology and Evolution, 38_(10), 4647-4654. https://doi.org/10.1093/molbev/msab199
  
- **CARD: Comprehensive Antibiotic Resistance Database**
  
   [Webpages](https://card.mcmaster.ca/)

  > Alcock, B. P., Huynh, W., Chalil, R., Smith, K. W., Raphenya, A. R., Wlodarski, M. A., ... & McArthur, A. G. (2023). CARD 2023: expanded curation, support for machine learning, and resistome prediction at the Comprehensive Antibiotic Resistance Database. _Nucleic Acids Research, 51_(D1), D690-D699. https://doi.org/10.1093/nar/gkac920

- **CheckM**
  
  [Code Repository](https://github.com/Ecogenomics/CheckM) | [Webpages](https://ecogenomics.github.io/CheckM/)

  > Parks, D. H., Imelfort, M., Skennerton, C. T., Hugenholtz, P., & Tyson, G. W. (2015). CheckM: assessing the quality of microbial genomes recovered from isolates, single cells, and metagenomes. _Genome Research, 25_(7), 1043-1055. http://www.genome.org/cgi/doi/10.1101/gr.186072.114.

- **chopper**
  
  [Code Repository](https://github.com/wdecoster/chopper)

  > De Coster, W., & Rademakers, R. (2023). NanoPack2: population-scale evaluation of long-read sequencing data. _Bioinformatics, 39_(5), btad311. https://doi.org/10.1093/bioinformatics/btad311

- **FastQC**
  
  [Code Repository](https://github.com/s-andrews/FastQC) | [Webpages](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)

  > Andrews, S. (2010). FastQC: A quality control tool for high throughput sequence data. Available online at: http://www.bioinformatics.babraham.ac.uk/projects/fastqc/

- **GTDB: Genome Taxonomy Database**

  [Webpages](https://gtdb.ecogenomic.org/)

  > Parks, D. H., Chuvochina, M., Rinke, C., Mussig, A. J., Chaumeil, P. A., & Hugenholtz, P. (2022). GTDB: an ongoing census of bacterial and archaeal diversity through a phylogenetically consistent, rank normalized and complete genome-based taxonomy. _Nucleic Acids Research, 50_(D1), D785-D794. https://doi.org/10.1093/nar/gkab776
  
  > Parks, D. H., Chuvochina, M., Chaumeil, P. A., Rinke, C., Mussig, A. J., & Hugenholtz, P. (2020). A complete domain-to-species taxonomy for Bacteria and Archaea. _Nature Biotechnology, 38_(9), 1079-1086. https://doi.org/10.1038/s41587-020-0501-8
  
  > Parks, D. H., Chuvochina, M., Waite, D. W., Rinke, C., Skarshewski, A., Chaumeil, P. A., & Hugenholtz, P. (2018). A standardized bacterial taxonomy based on genome phylogeny substantially revises the tree of life. _Nature Biotechnology, 36_(10), 996-1004. https://doi.org/10.1038/nbt.4229

  > Rinke, C., Chuvochina, M., Mussig, A. J., Chaumeil, P. A., Davín, A. A., Waite, D. W., ... & Hugenholtz, P. (2021). A standardized archaeal taxonomy for the Genome Taxonomy Database. _Nature Microbiology, 6_(7), 946-959. https://doi.org/10.1038/s41564-021-00918-8

- **GTDB-Tk: Genome Taxonomy Database and associated taxonomic classification toolkit**
  
  [Code Repository](https://github.com/Ecogenomics/GTDBTk) | [Webpages](https://ecogenomics.github.io/GTDBTk/index.html)

  > Chaumeil, P. A., Mussig, A. J., Hugenholtz, P., & Parks, D. H. (2022). GTDB-Tk v2: memory friendly classification with the genome taxonomy database. _Bioinformatics, 38_(23), 5315-5316. https://doi.org/10.1093/bioinformatics/btac672
  
  > Chaumeil, P. A., Mussig, A. J., Hugenholtz, P., & Parks, D. H. (2020). GTDB-Tk: a toolkit to classify genomes with the Genome Taxonomy Database. _Bioinformatics, 36_(6), 1925-1927. https://doi.org/10.1093/bioinformatics/btz848

- **Kracken 2**
  
  [Code Repository](https://github.com/DerrickWood/kraken2) | [Webpages](https://ccb.jhu.edu/software/kraken2/)

  > Lu, J., Rincon, N., Wood, D. E., Breitwieser, F. P., Pockrandt, C., Langmead, B., ... & Steinegger, M. (2022). Metagenome analysis using the Kraken software suite. _Nature Protocols, 17_(12), 2815-2839. https://doi.org/10.1038/s41596-022-00738-y
  
  > Wood, D. E., Lu, J., & Langmead, B. (2019). Improved metagenomic analysis with Kraken 2. _Genome Biology, 20_, 1-13. https://doi.org/10.1186/s13059-019-1891-0

- **LRBinner**
  
  [Code Repository](https://github.com/anuradhawick/LRBinner)

  > Wickramarachchi, A., & Lin, Y. (2021). LRBinner: binning long reads in metagenomics datasets. In _21st International Workshop on Algorithms in Bioinformatics (WABI 2021): Leibniz International Proceedings in Informatics (LIPIcs) 201_, 11:1-11:18. https://10.4230/LIPIcs.WABI.2021.11
  
  > Wickramarachchi, A., & Lin, Y. (2022). Binning long reads in metagenomics datasets using composition and coverage information. _Algorithms for Molecular Biology, 17_(1), 14. https://doi.org/10.1186/s13015-022-00221-z

- **metaFlye**
  
  [Code Repository](https://github.com/mikolmogorov/Flye)

  > Kolmogorov, M., Bickhart, D. M., Behsaz, B., Gurevich, A., Rayko, M., Shin, S. B., ... & Pevzner, P. A. (2020). metaFlye: scalable long-read metagenome assembly using repeat graphs. _Nature Methods, 17_(11), 1103-1110. https://doi.org/10.1038/s41592-020-00971-x

- **minimap2**
  
  [Code Repository](https://github.com/lh3/minimap2) | [Webpages](https://lh3.github.io/minimap2/)

  > Li, H. (2021). New strategies to improve minimap2 alignment accuracy. _Bioinformatics, 37_(23), 4572-4574. https://doi.org/10.1093/bioinformatics/btab705
  
  > Li, H. (2018). Minimap2: pairwise alignment for nucleotide sequences. _Bioinformatics, 34_(18), 3094-3100. https://doi.org/10.1093/bioinformatics/bty191

- **MultiQC**
  
  [Code Repository](https://github.com/MultiQC/MultiQC) | [Webpages](https://multiqc.info/)

  > Ewels, P., Magnusson, M., Lundin, S., & Käller, M. (2016). MultiQC: summarize analysis results for multiple tools and samples in a single report. _Bioinformatics, 32_(19), 3047-3048. https://doi.org/10.1093/bioinformatics/btw354

- **Plasmer**
  
  [Code Repository](https://github.com/nekokoe/Plasmer) 

  > Zhu, Q., Gao, S., Xiao, B., He, Z., & Hu, S. (2023). Plasmer: an accurate and sensitive bacterial plasmid prediction Tool Based on Machine Learning of Shared k-mers and genomic features. _Microbiology Spectrum, 11_(3), e04645-22. https://doi.org/10.1128/spectrum.04645-22

- **Porechop**
  
  [Code Repository](https://github.com/rrwick/Porechop)

  > Wick, R. R., Judd, L. M., Gorrie, C. L., & Holt, K. E. (2017). Completing bacterial genome assemblies with multiplex MinION sequencing. Microbial genomics, 3(10), e000132. https://doi.org/10.1099/mgen.0.000132

- **QUAST: Quality Assessment Tool for Genome Assemblies**

  [Code Repository](https://github.com/ablab/quast) | [Webpages](https://quast.sourceforge.net/)

  > Gurevich, A., Saveliev, V., Vyahhi, N., & Tesler, G. (2013). QUAST: quality assessment tool for genome assemblies. _Bioinformatics, 29_(8), 1072-1075. https://doi.org/10.1093/bioinformatics/btt086

- **ResFinder**
  
  [Code Repository](https://github.com/cadms/resfinder) | [Webpages](https://www.genomicepidemiology.org/services/)

  > Bortolaia, V., Kaas, R. S., Ruppe, E., Roberts, M. C., Schwarz, S., Cattoir, V., ... & Aarestrup, F. M. (2020). ResFinder 4.0 for predictions of phenotypes from genotypes. _Journal of Antimicrobial Chemotherapy, 75_(12), 3491-3500. https://doi.org/10.1093/jac/dkaa345

- **RGI: Resistance Gene Identifier**
  
  [Code Repository](https://github.com/arpcard/rgi) | [Webpages](https://card.mcmaster.ca/analyze/rgi)

  > Alcock, B. P., Huynh, W., Chalil, R., Smith, K. W., Raphenya, A. R., Wlodarski, M. A., ... & McArthur, A. G. (2023). CARD 2023: expanded curation, support for machine learning, and resistome prediction at the Comprehensive Antibiotic Resistance Database. _Nucleic Acids Research, 51_(D1), D690-D699. https://doi.org/10.1093/nar/gkac920

- **SAMtools**
  
  [Code Repository](https://github.com/samtools/samtools) | [Webpages](https://www.htslib.org/)

  > Danecek, P., Bonfield, J. K., Liddle, J., Marshall, J., Ohan, V., Pollard, M. O., ... & Li, H. (2021). Twelve years of SAMtools and BCFtools. _Gigascience, 10_(2), giab008. https://doi.org/10.1093/gigascience/giab008

- **seqtk**
  
  [Code Repository](https://github.com/lh3/seqtk) | [Webpages](https://docs.csc.fi/apps/seqtk/)

  > Li, H. (2023). seqtk-1.4 [Computer Software]. Retrieved from https://github.com/lh3/seqtk

## Python Packages
- **Python**

  [Webpages](https://www.python.org/)

  > Van Rossum,  G., & Drake, F. L. (2009). _Python 3 Reference Manual_. Scotts Valley, CA: CreateSpace.
  
- **Biopython**

  [Code Repository](https://github.com/biopython/biopython) | [Webpages](https://biopython.org/)
  
  > Cock, P. J., Antao, T., Chang, J. T., Chapman, B. A., Cox, C. J., Dalke, A., ... & De Hoon, M. J. (2009). Biopython: freely available Python tools for computational molecular biology and bioinformatics. _Bioinformatics, 25_(11), 1422. https://doi.org/10.1093%2Fbioinformatics%2Fbtp163
  
- **NumPy**

  [Code Repository](https://github.com/numpy/numpy) | [Webpages](https://numpy.org/)

  > Harris, C. R., Millman, K. J., Van Der Walt, S. J., Gommers, R., Virtanen, P., Cournapeau, D., ... & Oliphant, T. E. (2020). Array programming with NumPy. _Nature, 585_(7825), 357-362. https://doi.org/10.1038/s41586-020-2649-2
  
- **pandas**
  
  [Code Repository](https://github.com/pandas-dev/pandas) | [Webpages](https://pandas.pydata.org/)
  
  > McKinney, W. (2011). pandas: a foundational Python library for data analysis and statistics. _Python for High Performance and Scientific Computing, 14_(9), 1-9. https://www.researchgate.net/publication/265194455
  
- **seaborn**
  
  [Code Repository](https://github.com/mwaskom/seaborn) | [Webpages](https://seaborn.pydata.org/)
  
  > Waskom, M. L. (2021). Seaborn: statistical data visualization. _Journal of Open Source Software, 6_(60), 3021. https://doi.org/10.21105/joss.03021


## Software Packaging/Containerisation Tools

- **Anaconda**

  [Webpages](https://anaconda.com)

  > _Anaconda Software Distribution._ (2020). _Anaconda Documentation_. Anaconda Inc. Retrieved from https://docs.anaconda.com/

- **Apptainer/Singularity**

  [Code Repository](https://github.com/apptainer/apptainer) | [Webpages](https://apptainer.org/)

  > Singularity Developers. (2020). Singularity. 10.5281/zenodo.1310023. https://doi.org/10.5281/zenodo.1310023

  > Kurtzer, G. M., Sochat, V., & Bauer, M. W. (2017). Singularity: Scientific containers for mobility of compute. PloS one, 12(5), e0177459. https://doi.org/10.1371/journal.pone.0177459
  
- **Bioconda**

  [Code Repository](https://github.com/bioconda) | [Webpages](https://bioconda.github.io/)

  > Grüning, B., Dale, R., Sjödin, A., Chapman, B. A., Rowe, J., Tomkins-Tinch, C. H., ... & Bioconda Team. (2018). Bioconda: sustainable and comprehensive software distribution for the life sciences. _Nature Methods, 15_(7), 475-476. https://doi.org/10.1038/s41592-018-0046-7

- **BioContainers**
  
  [Code Repository](https://github.com/BioContainers) | [Webpages](https://biocontainers.pro/)

  > da Veiga Leprevost, F., Grüning, B. A., Alves Aflitos, S., Röst, H. L., Uszkoreit, J., Barsnes, H., ... & Perez-Riverol, Y. (2017). BioContainers: an open-source and community-driven framework for software standardization. _Bioinformatics, 33_(16), 2580-2582. https://doi.org/10.1093/bioinformatics/btx192

- **Docker**
  
  [Code Repository](https://github.com/docker) | [Webpages](https://www.docker.com/)

  > Merkel, D. (2014). Docker: lightweight linux containers for consistent development and deployment. _Linux Journal, 239_(2), 2. Retrieved from https://www.linuxjournal.com/content/docker-lightweight-linux-containers-consistent-development-and-deployment
