{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "bf620a33-7c57-472b-84ad-46bffcaa8a36",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Required modules\n",
    "import gzip\n",
    "import os\n",
    "import sys\n",
    "import shutil"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0b6bb076-b829-4bf6-bc61-8d11eacdc10d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parsing arguments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "16cbff01-ad2b-455d-8681-670b63944f39",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Making a notebook to standardize seqeuncing data and refernce genomes. Will late be switched to a script."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "51ae3b1b-27fd-4373-b786-cb11ac0ad24b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def decompress_fastq(fastq_input, sample_name):\n",
    "    \"\"\"\n",
    "    This function takes a fastq input (either gzipped or no gzipped) and\n",
    "    uncompresses it to a temporary file. If its already gzipped it just \n",
    "    makes a copy to the temp file instead. It will name the temp file \n",
    "    according to the sample_name string\n",
    "\n",
    "    This returs the file name of the uncompressed fastq\n",
    "    \"\"\"\n",
    "    # Determining if the file is in gzip format\n",
    "    temp_uncompressed = sample_name + '_unc_tmp.fastq' #temp_file_name\n",
    "    \n",
    "    try:\n",
    "        with open(fastq_input, 'rb') as f:\n",
    "            # Read the first two bytes\n",
    "            magic_number = f.read(2)\n",
    "        \n",
    "        # Check if the magic number indicates gzip format\n",
    "        if magic_number == b'\\x1f\\x8b':\n",
    "            # Decompress a file\n",
    "            with gzip.open(fastq_input, 'rb') as f_in:\n",
    "                with open(temp_uncompressed, 'wb') as f_out:\n",
    "                    f_out.writelines(f_in)\n",
    "            print('File ' + fastq_input + ' was unzipped and written to temporary file ' + temp_uncompressed)        \n",
    "        \n",
    "        # assume the file is plain text if not gzipped\n",
    "        else:\n",
    "            shutil.copy(fastq_input, temp_uncompressed)\n",
    "            print('File ' + fastq_input + ' was already unzipped and written to temporary file ' + temp_uncompressed)\n",
    "    \n",
    "    except Exception as e:\n",
    "            print(f\"Error processing {fastq_input}: {e}\", file=sys.stderr)\n",
    "\n",
    "    return(temp_uncompressed)\n",
    "\n",
    "\n",
    "def check_fastq(fastq):\n",
    "    \"\"\"\n",
    "    This function checks to make sure a fastq file seems to be in the correct format.\n",
    "\n",
    "    This function expects fastqs to be uncompressed\n",
    "\n",
    "    This function returns a True if its in the correct format.\n",
    "    \"\"\"\n",
    "    \n",
    "    phred = {}\n",
    "    for x in range(0, 94):\n",
    "        phred[x] = chr(x + 33)\n",
    "    \n",
    "    \n",
    "    # Determinging if the file is in standard fastq format\n",
    "    try:\n",
    "        with open(fastq, 'r') as f:\n",
    "            valid_format = True\n",
    "            line_num = 0\n",
    "            for line in f:\n",
    "                line = line.strip()\n",
    "                line_num += 1\n",
    "                \n",
    "                if line_num % 4 == 1:\n",
    "                    if not line.startswith('@'):\n",
    "                        valid_format = False\n",
    "                        raise ValueError(\"File Format Error Line \" + str(line_num) + ' Does Not Start with a @')\n",
    "                        \n",
    "                elif line_num % 4 == 2:\n",
    "                    # sequence line, check for valid characters\n",
    "                    for char in line:\n",
    "                        if char not in 'ACGTN':\n",
    "                            valid_format = False\n",
    "                            raise ValueError(\"File Format Error Line \" + str(line_num) + ' Contains Charcters other than ATGCN')\n",
    "                            \n",
    "                elif line_num % 4 == 3:\n",
    "                    if not line.startswith('+'):\n",
    "                        valid_format = False\n",
    "                        raise ValueError(\"File Format Error Line \" + str(line_num) + ' Does Not Start with a +')\n",
    "                        \n",
    "                elif line_num % 4 == 0:\n",
    "                    for char in line:\n",
    "                        if not char in phred.values():\n",
    "                            valid_format = False\n",
    "                            print(valid_format)\n",
    "                            raise ValueError(\"File Format Error Line \" + str(line_num) + ' Contains a non-standard phred score')\n",
    "                            \n",
    "        \n",
    "            # Check if the number of lines is a multiple of 4 (valid FASTQ format)\n",
    "            if line_num % 4 != 0:\n",
    "                raise ValueError(\"File Format Error: FastQ had \" + str(line_num) + ' lines, not a multiple of 4')\n",
    "                valid_format = False\n",
    "            \n",
    "    except Exception as e:\n",
    "        print(f\"Error processing {fastq}: {e}\", file=sys.stderr)\n",
    "\n",
    "    if valid_format:\n",
    "        print(fastq + ' is a proper fasta file')\n",
    "        return True\n",
    "    else:\n",
    "        print(fastq + ' is NOT a proper fasta file')\n",
    "        return False\n",
    "\n",
    "\n",
    "\n",
    "def rename_fastqs(in_fq, out_fq, prefix):\n",
    "    \"\"\"\n",
    "    This function takes a fastq file (Uncompressed) and renames\n",
    "    the sequences in the file. This function is used to rename\n",
    "    complex sequene names typical of ONT sequencing. By defualt\n",
    "    this function renames sequences using the supplied prefix and\n",
    "    the read count after that (\"{prefix}_{read#}\")\n",
    "    \"\"\"\n",
    "    \n",
    "    try:\n",
    "        with open(in_fq, 'r') as infile, open(out_fq, 'w') as outfile:\n",
    "            line_num = 0\n",
    "            read_count = 0\n",
    "            for line in infile:\n",
    "                line = line.strip()\n",
    "                if line_num % 4 == 0:\n",
    "                    # Sequence identifier line (starts with '@')\n",
    "                    read_count += 1\n",
    "                    sequence_name = f\"{prefix}_{read_count}\"\n",
    "                    outfile.write(f\"@{sequence_name}\\n\")\n",
    "                else:\n",
    "                    # Other lines (sequence, optional quality scores)\n",
    "                    outfile.write(f\"{line}\\n\")\n",
    "                line_num += 1\n",
    "            print(f\"Successfully renamed sequences, written to {out_fq}\")\n",
    "            return out_fq\n",
    "    except Exception as e:\n",
    "        print(f\"Error reading or writing files: {e}\", file=sys.stderr)\n",
    "    \n",
    "\n",
    "    \n",
    "def gzip_fq(in_fq, out_fq):\n",
    "    \"\"\"\n",
    "    Takes a uncompressed fastq and gzips it.\n",
    "    \"\"\"\n",
    "\n",
    "    try:\n",
    "        with open(in_fq, 'rb') as f_in, gzip.open(out_fq, 'wb') as f_out:\n",
    "            shutil.copyfileobj(f_in, f_out)\n",
    "        \n",
    "        print(f\"Successfully compressed {in_fq} to {out_fq}\")\n",
    "        \n",
    "    except Exception as e:\n",
    "        print(f\"Error compressing file: {e}\", file=sys.stderr)\n",
    "\n",
    "def clean_files(file_list):\n",
    "    \"\"\"\n",
    "    cleans up temp files\n",
    "    \"\"\"\n",
    "\n",
    "    try:\n",
    "        for file in file_list:\n",
    "            os.remove(file)\n",
    "            print(f\"Successfully deleted {file}\")\n",
    "        \n",
    "    except Exception as e:\n",
    "        print(f\"Error deleting file: {e}\", file=sys.stderr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "1e9b16e7-5ad8-468f-8b29-22e75a993b29",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Temp inputs for testing\n",
    "fastq_input = 'test_notebook/barcode01_subset.fastq.gz'\n",
    "\n",
    "sample_name = 'sample_1'\n",
    "\n",
    "rename = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "d4b07944-4a52-47ec-8eb0-259006c70f4c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File test_notebook/barcode01_subset.fastq.gz was unzipped and written to temporary file sample_1_unc_tmp.fastq\n",
      "sample_1_unc_tmp.fastq is a proper fasta file\n",
      "Successfully renamed sequences, written to output_uncompressed\n",
      "Successfully compressed renammed.fastq to sample_1.fastq.gz\n",
      "Successfully deleted sample_1_unc_tmp.fastq\n",
      "Successfully deleted renammed.fastq\n",
      "fastq for sample_1 was checked and compressed. It can be found at sample_1.fastq.gz\n"
     ]
    }
   ],
   "source": [
    "temp_files = [] #Creating list of temp files to clean up\n",
    "\n",
    "decompressed = decompress_fastq(fastq_input, sample_name) #Decompressin (or not if it already is) input file and renamming it\n",
    "temp_files.append(decompressed) #adding temp file for later deltion\n",
    "\n",
    "if check_fastq(decompressed): # Checking if it is a standard fastq\n",
    "    if rename:\n",
    "        decompressed = rename_fastqs(decompressed, 'renammed.fastq', sample_name) # renamming sequences if specified\n",
    "        temp_files.append(decompressed) # adding renammed seq file to temp files to be deleted\n",
    "        \n",
    "    output_sample = sample_name + '.fastq.gz' # creating file file name based on sample name\n",
    "    \n",
    "    gzip_fq(decompressed, output_sample) # gziping final file\n",
    "    \n",
    "    clean_files(temp_files) # cleaning temp files\n",
    "    \n",
    "    print('fastq for ' + sample_name + ' was checked and compressed. It can be found at ' + output_sample)\n",
    "    \n",
    "else:\n",
    "    print('Sample ' + sample_name + ' did not pass fastq check. See errors above')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd428e0b-806d-43d1-b0bc-c5df963f0299",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
